### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by doemefu**
This pull request documents two important feedback items in the project memory, clarifying security and JWT scope handling edge cases in Spring Security 7 and Spring Authorization Server 7. It also records the completion of the IoT Device OAuth2 Clients feature, including implementation details, operational impacts, and follow-up actions.

**New feedback and documentation:**

*Method Security and Authorization Handling:*
- Added `.claude/memory/feedback_method_security_required.md` to explain that `@PreAuthorize` is ignored unless `@EnableMethodSecurity` is present, and that `AuthorizationDeniedException` must be mapped to 403 in the global exception handler to avoid silent 500 errors. Also recommends adding explicit tests for insufficient authority cases.

*JWT Scope Claim Shape:*
- Added `.claude/memory/feedback_ss7_jwt_scope_shape.md` documenting that Spring Authorization Server 7 emits the `scope` claim as a JSON array in JWTs but as a space-separated string in token responses, and provides guidance for writing compatible integration tests.

**Project memory and release documentation:**

*IoT Device OAuth2 Clients Feature:*
- Updated `.claude/memory/MEMORY.md` with a detailed entry for the IoT Device OAuth2 Clients implementation, including migration to JDBC `RegisteredClientRepository`, new admin API, JWT claim extensions, caching strategy, and operational runbook changes. Outlines completed tests, plan reviewer findings, and pending operational follow-ups.

## Summary by CodeRabbit

* **New Features**
  * Added device OAuth2 client management API (`/api/v1/clients`) to create, list, and delete IoT device clients.
  * Device client credentials now include a `device_id` JWT claim for MQTT authentication.
  * Persistent JDBC-backed OAuth2 client storage with YAML-seeded SSO clients on first boot.

* **Documentation**
  * Updated README with Device Client Admin API details and endpoint descriptions.
  * Added operational runbook for device client lifecycle and secret rotation.
  * Added IoT device client specification.

* **Configuration**
  * New device client configuration section with access token TTL and allowed scopes.

[Review Change Stack](https://app.coderabbit.ai/change-stack/doemefu/homelab-auth-service/pull/45)

### CODE REVIEW

File: src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+        jdbcTemplate.update(
+                "DELETE FROM oauth2_authorization WHERE registered_client_id = ?", id);
+        jdbcTemplate.update(
+                "DELETE FROM oauth2_authorization_consent WHERE registered_client_id = ?", id);
+        jdbcTemplate.update(
+                "DELETE FROM oauth2_registered_client WHERE id = ?", id);
+        clientKindLookup.invalidate(id);
+        log.info("Deleted device client '{}'", clientId);
+    }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java:162]: Intentional — the idempotent-delete design is documented in the method Javadoc. The `client_kind='device'` guard prevents accidental deletion of SSO clients; returning 204 for both cases keeps the caller's logic simple and mirrors a standard REST idempotent DELETE. A WARN log was considered but would generate noise on any mis-typed client ID; the guard is the safety net.

File: src/main/java/ch/furchert/homelab/auth/service/ClientKindLookup.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private final JdbcTemplate jdbcTemplate;
+    private final ConcurrentHashMap<String, String> cache = new ConcurrentHashMap<>();
+
+    /**
+     * @param registeredClientId the PK from oauth2_registered_client.id
+     * @return the client_kind value, or null if the row does not exist
+     */
+    public String lookup(String registeredClientId) {
+        return cache.computeIfAbsent(registeredClientId, this::loadFromDb);
+    }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/service/ClientKindLookup.java:36]: Not addressing. The cache is bounded by the number of registered clients (a handful in this homelab), so unbounded growth is not a concern. The null-caching gap (unknown IDs re-querying on every call) is also not actionable: unknown IDs in production would indicate a token for a deleted client, which is a rare edge case. The default-deny behaviour (missing kind → no device_id claim) is correct.

File: src/test/java/ch/furchert/homelab/auth/AbstractIntegrationTest.java [hunk 3]
```diff
@@ -14,7 +14,7 @@ public abstract class AbstractIntegrationTest {
     // Do NOT use @Testcontainers + @Container here: JUnit 5's AfterAllCallback
     // stops @Container static fields after each concrete subclass finishes, which
     // kills the container before the next integration test class can connect.
-    static final PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:17-alpine");
+    static final PostgreSQLContainer postgres = new PostgreSQLContainer("postgres:17-alpine");
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/test/java/ch/furchert/homelab/auth/AbstractIntegrationTest.java:17]: Not a raw type. The new import `org.testcontainers.postgresql.PostgreSQLContainer` (Testcontainers 2.x) is a concrete, non-generic class — it extends `JdbcDatabaseContainer<PostgreSQLContainer>` and carries no open type parameter. The `<?\>` suggestion applies to the old 1.x self-typed API (`PostgreSQLContainer<SELF extends ...>`); it does not apply here. The import and the type are both correct.

File: src/main/java/ch/furchert/homelab/auth/controller/DeviceClientController.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * </ul>
+ * <p>
+ * Security note: the create() response body contains a plaintext client secret
+ * returned exactly once. Do not log the response body. Sentry is configured
+ * (application.yaml) without HTTP body capture, so default breadcrumbs are safe.
+ */
+@RestController
+@RequestMapping("/api/v1/clients")
+@RequiredArgsConstructor
+@PreAuthorize("hasRole('ADMIN') or hasAuthority('SCOPE_clients:admin')")
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/controller/DeviceClientController.java:31]: Not addressing in this PR. The suggestion is valid as a future hardening item but it is a feature addition (a new test class with a dedicated integration context), not a bug in the current implementation. The existing `DeviceClientLifecycleIT` exercises the authorized path. Tracked separately.

File: src/main/java/ch/furchert/homelab/auth/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -40,6 +42,17 @@ public ResponseEntity<Map<String, Object>> handleBadCredentials(BadCredentialsEx
...
+    /**
+     * @PreAuthorize failures raise AuthorizationDeniedException; legacy access
+     * decisions raise AccessDeniedException. Both must return 403 — otherwise
+     * the catch-all generic handler below maps them to 500.
+     */
+    @ExceptionHandler({AuthorizationDeniedException.class, AccessDeniedException.class})
+    public ResponseEntity<Map<String, Object>> handleAuthorizationDenied(RuntimeException ex) {
+        return ResponseEntity.status(HttpStatus.FORBIDDEN)
+                .body(errorBody(HttpStatus.FORBIDDEN, "Forbidden", null));
+    }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/exception/GlobalExceptionHandler.java:54]: Not addressing. The `ex` parameter is required in the method signature for Spring to dispatch `@ExceptionHandler` correctly — it cannot be removed. Intentionally not using it in the body to avoid leaking authorization internals in the 403 response. On the redundancy point: `AuthorizationDeniedException extends AccessDeniedException` is correct, but listing both explicitly documents which concrete types are expected here; the redundancy is intentional for clarity.

File: src/main/java/ch/furchert/homelab/auth/dto/DeviceClientCreatedResponse.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Returned from POST /api/v1/clients. Contains the plaintext client secret —
+ * shown once and never persisted in clear. Never log this response body.
+ */
+public record DeviceClientCreatedResponse(
+        String clientId,
+        String clientSecret,
+        List<String> scopes,
+        Instant createdAt
+) {
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/dto/DeviceClientCreatedResponse.java:15]: Not addressing. Spring's default access-log does not capture response bodies, and Sentry is configured in `application.yaml` to exclude HTTP body capture. Adding a custom `toString` override or a logging filter is over-engineering for this risk level. The existing Javadoc warning on the DTO is sufficient.

File: src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        String id = UUID.randomUUID().toString();
+        Instant now = Instant.now();
+        RegisteredClient client = RegisteredClient.withId(id)
+                .clientId(clientId)
+                .clientIdIssuedAt(now)
+                .clientName(description == null ? clientId : description)
+                .clientSecret(hashedSecret)
+                .clientAuthenticationMethod(ClientAuthenticationMethod.CLIENT_SECRET_BASIC)
+                .authorizationGrantType(AuthorizationGrantType.CLIENT_CREDENTIALS)
+                .scopes(s -> s.addAll(scopes))
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java:83]: Not addressing in this PR. All device clients currently receive the same scope set from `allowed-scopes` in config, which is the intended behaviour for the initial implementation. Allowing per-client scope selection at creation time is a design change that requires updating the request DTO, validation logic, and spec — tracked as a follow-up.

File: src/main/java/ch/furchert/homelab/auth/config/SecurityConfig.java [hunk 8]
```diff
@@ -100,12 +112,29 @@ public PasswordEncoder passwordEncoder() {
...
-        converter.setJwtGrantedAuthoritiesConverter(authoritiesConverter);
+        converter.setJwtGrantedAuthoritiesConverter(jwt -> {
+            Collection<GrantedAuthority> authorities = new ArrayList<>();
+            Collection<GrantedAuthority> scopes = scopesConverter.convert(jwt);
+            Collection<GrantedAuthority> roles = rolesConverter.convert(jwt);
+            authorities.addAll(scopes);
+            authorities.addAll(roles);
+            return authorities;
+        });
         return converter;
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/config/SecurityConfig.java:138]: Not a real issue. Verified from the Spring Security 7.0.5 source (`spring-security-oauth2-resource-server-7.0.5-sources.jar`): `JwtGrantedAuthoritiesConverter.convert()` always returns a `new ArrayList<>()` — it never returns null, even when the targeted claim is absent. No null guards are needed. The Javadoc comment in SecurityConfig was corrected to accurately reflect this.

File: src/main/java/ch/furchert/homelab/auth/config/StaticClientSeeder.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        int seeded = 0;
+        for (OidcClientProperties.ClientDefinition def : properties.getClients()) {
+            if (registeredClientRepository.findByClientId(def.getClientId()) != null) {
+                log.debug("SSO client '{}' already present in DB; skipping seed", def.getClientId());
+                continue;
+            }
+            RegisteredClient client = buildRegisteredClient(def);
+            registeredClientRepository.save(client);
+            seeded++;
+            log.info("Seeded SSO client '{}' (id={})", def.getClientId(), client.getId());
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/config/StaticClientSeeder.java:53]: Not addressing. This is a single-instance homelab deployment; concurrent startup of two auth-service pods is not a scenario. The race window at startup is milliseconds, and the consequence (a unique-key violation on a known-idempotent operation) would simply crash startup — detectable immediately, not a silent failure.

File: src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    private final JdbcTemplate jdbcTemplate;
+    private final PasswordEncoder passwordEncoder;
+    private final ClientKindLookup clientKindLookup;
+    private final OidcClientProperties properties;
+
+    @Transactional
+    public DeviceClientCreatedResponse create(String clientId, String description) {
+        if (registeredClientRepository.findByClientId(clientId) != null) {
+            throw new ResourceConflictException("Client '" + clientId + "' already exists");
+        }
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/service/DeviceClientService.java:58]: Not addressing for the same reason as StaticClientSeeder: single-instance deployment. The `@Transactional` boundary provides isolation for same-process concurrent calls, and concurrent distinct processes are not a scenario in this homelab.

File: src/main/java/ch/furchert/homelab/auth/service/ClientKindLookup.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    private String loadFromDb(String registeredClientId) {
+        try {
+            return jdbcTemplate.queryForObject(
+                    "SELECT client_kind FROM oauth2_registered_client WHERE id = ?",
+                    String.class,
+                    registeredClientId);
+        } catch (EmptyResultDataAccessException _) {
+            return null;
```
[the remark this answers is not in the document]
Comment (doemefu) [reply src/main/java/ch/furchert/homelab/auth/service/ClientKindLookup.java:53]: Not a real issue. Unnamed variables (`_`) were standardised in Java 22 (JEP 456) and are valid in catch clauses. The project targets Java 25, which fully supports this syntax. The code compiles without error or warning. CodeRabbit appears to be applying Java 21 rules (where `_` was made a reserved keyword as an *identifier*, not as an unnamed variable).
